File: frontend/src/services/donationService.ts
import API from "../api/axios";

export interface DonationItem {

  item_name: string;

  quantity: number;

  confidence: number;

  condition: string;

}

export interface DonationPayload {

  donor_id: number;

  latitude: number;

  longitude: number;

  items: DonationItem[];

}

export const submitDonation = async (
  data: DonationPayload
) => {

  const response = await API.post(
    "/donations/",
    data
  );

  return response.data;

};

File: backend/app/services/donation_service.py
from sqlalchemy.orm import Session

from app.models.donation import Donation
from app.models.donation_item import DonationItem

from app.schemas.donation_schema import DonationCreate

from app.services.matching_service import generate_matches


def create_donation(
    data: DonationCreate,
    db: Session,
):

    donation = Donation(

        donor_id=data.donor_id,

        latitude=data.latitude,

        longitude=data.longitude,

    )

    db.add(donation)
    db.commit()
    db.refresh(donation)

    for item in data.items:

        donation_item = DonationItem(

            donation_id=donation.id,

            item_name=item.item_name,

            quantity=item.quantity,

            confidence=item.confidence,

            condition=item.condition,

        )

        db.add(donation_item)

    db.commit()

    generate_matches(
        donation.id,
        db,
    )

    return {

        "id": donation.id,

        "donor_id": donation.donor_id,

        "status": donation.status,

        "message": "Donation submitted successfully. Matching completed.",

    }